mlflow/types/schema.py

```diff
@@ -645,23 +658,52 @@ def _merge(self, map_type: "Map") -> "Map":
                 f"{map_type} with value_type={map_type.value_type}"
             )
 
-        if (
-            isinstance(self.value_type, (Array, Object, Map))
-            and self.value_type.__class__ is map_type.value_type.__class__
-        ):
+        if isinstance(self.value_type, (Array, Object, Map, AnyType)):
             return Map(value_type=self.value_type._merge(map_type.value_type))
 
         raise MlflowException(f"Map type {self!r} and {map_type!r} are incompatible")
 
 
+class AnyType:
+    def __init__(self):
+        """
+        AnyType can store any json-serializable data including None values.
+
+        .. Note::
+            AnyType should not be used to host union types in a schema.
+            It should only be used when the field is None, the type is not known
+            at the time of data creation, or the field can have multiple types.
+            e.g. for GenAI flavors, the model output could contain `None` values,
+            and `AnyType` can be used to represent them.
```

Comment:
what's the issue with using it for union types?

Reply:
Do we have a plan to introduce "Optional" type? In the case of potential None, it may be more helpful if we could specify the type with Optional[long] than using "Any".

Reply:
Sorry if this is already discussed, but did we consider introducing Option type? If the main aim of AnyType is supporting `None`, it may be an option too?

Reply:
Since some types contain required field, introducing Option type is a bit confusing, and in reality AnyType can host any type not just None. Updated the docstring to make it less confusing

Reply:
Yea agree that we don't need `Optional` type. The edge case Tomu raised is still valid, maybe we can raise when  `AnyType` is used with `required=True`? User might imagine "Any type but None" but the enforcement system doesn't work in that way.

Reply:
AnyType with required=True means that the field has to exist, and the value can be None. While AnyType with required=False means the field doesn't need to exist
e.g. `[{"a": None, "b": 123}, {"b": 234}]` here "a" should be optional AnyType
`[{"a": None}]` here "a" should be required AnyType

Reply:
Oooh I thought `required=True` doesn't allow `None`. It's quite confusing imo because the way we make the field "optional" is passing `None`.... then I guess it's natural to assume "required" is not for `None`...

> In order to configure a field as optional, you must pass in a value of None or np.nan for that field when using the [mlflow.models.infer_signature()](https://mlflow.org/docs/latest/python_api/mlflow.models.html#mlflow.models.infer_signature) function.

Reply:
e.g. `[{"a": None}, {"a": "abc"}]` will still be inferred as String (optional).
Yea I'm a bit hesitating here as well, I noticed that GenAI model outputs contain fields whose values are None, but if we mark it as optional then this field can be omitted, I'm not sure if that's an error or not.

Reply:
Yea I don't think we need a change here but this is just another reason we should migrate to pydantic based schema. Current one seems to deviate much from Python type system...🫠